### PR TITLE
Show better error message for invalid operation to stream

### PR DESCRIPTION
**Pull request by demotomohiro**
Calling `setPosition`/`getPosition` to a pipe stream shows runtime error message "You cannot use setPosition to this Stream" rather than "SIGSEGV: Illegal storage access. (Attempt to readfrom nil?)".
Reading or peeking to `inputStream` in osproc module shows runtime error message "You cannot read/peek from this stream".
Writing to `outputStream`, `errorStream`, `peekableOutputStream`, and `peekableErrorStream` in osproc module shows runtime error message.
Writing to a file stream with `fmRead` and reading to a file stream with `fmWrite` shows runtime error message.

You cannot seek for pipe:
https://man7.org/linux/man-pages/man2/lseek.2.html

On windows, `Stream.setPositionImpl` and `Stream.getPositionImpl` of `inputStream`, `outputStream` and `errorStream` are nil.
As `setPosition(s: Stream, pos: int)` and `getPosition*(s: Stream): int` don't check nil, calling them to `inputStream` just cause SIGSEGV.
On other OS, `Stream.setPositionImpl` and `Stream.getPositionImpl` of `input/output/errorStream` are set while they are error.

Reading from `stdin` or writing to `stdout` of child process using pipe is meaningless.

Detecting these invalid procedure call to the pipe at compile time would be better, but I don't think that is possible with exising `Stream`.
It would requires generic Stream type like this:
```nim
type
  StreamOption* = enum
    soReadable,
    soWritable,
    soSeekable,
    soPeekable

  Stream*[SO: static[set[StreamOption]]] = ref StreamObj
  StreamObj = object

proc setPosition*[SO: static[set[StreamOption]]](s: Stream[SO], pos: int) =
  when soSeekable notin SO:
    {.fatal: "You cannot use setPosition to this stream".}
```

So set procedural types of `Stream` to nil if it should not be called and check if it is nil with `assert` when it is going to be called.

### CODE REVIEW

File: lib/pure/streams.nim [hunk 1]
```diff
@@ -135,6 +135,12 @@ type
     flushImpl*: proc (s: Stream)
       {.nimcall, raises: [Defect, IOError, OSError], tags: [WriteIOEffect], gcsafe.}
 
+  StreamOption* = enum  ## What operation the stream can do.
+    soReadable,         ## You can read data from stream.
+    soWritable,         ## Data can be written to the stream.
+    soSeekable,         ## Read or write position can be changed.
+    soPeekable          ## Can read data without advancing read position.
+
```
Comment (Araq) [inline lib/pure/streams.nim:143]: Why is this required? Shouldn't the new asserts be good enough?

Comment (demotomohiro) [reply lib/pure/streams.nim:143]: `proc newFileStream*(f: File)` is called from several procedures and it always set procedure pointers in stream even if it must not be called.
These procedure pointers need to be set to nil to assert check works.
`newFileStream` and `openFileStream` proc with filename parameter and `inputStream`/`outputStream`/`errorStream` procs in osproc module except windows call `proc newFileStream*(f: File)`.

`StreamOption` allows these procs to ask `proc newFileStream*(f: File; options: set[StreamOption])` to set invalid procedure pointers to nil.
It can also be used for creating a stream from unseekable file descripter (pipe, socket, FIFO, etc) that raise assertion when `setPosition`/`getPosition` is called.
Should these procs set invalid procedure pointers to nil after calling `proc newFileStream*(f: File)` instead of adding `options: set[StreamOption]` parameter to it?

**Comment by Araq** [comment]
Bad cost/benefit tradeoff -- lot's of code added just for improving the error message for some developers while every developer suffers from the added code a tiny bit at compile- and runtime. Sorry.
